File: leetcode/101-200/140-word-break-ii/bench/word_break_ii.c
#include <stdio.h>
#include <stdlib.h>

// Exponential-backtracking segmentation COUNT for LeetCode #140. Short strings /
// small dicts keep the search tractable; counts are bounded by compositions of
// SLEN (≈10^4) so the sink never overflows. The dict SET is a flat stamped
// base-A table (hand-rolled; see word_break_ii.kara header and #139).
static long alpha = 3, minlen = 1, maxlen = 3, slen = 16;
static long *s;
static char *table;
static long base[8];

static long count(long start) {
    if (start == slen) return 1;
    long total = 0, code = 0;
    for (long end = start + 1; end <= slen && end - start <= maxlen; end++) {
        code = code * alpha + s[end - 1];
        long ln = end - start;
        if (ln >= minlen) {
            if (table[base[ln] + code]) total += count(end);
        }
    }
    return total;
}

int main(void) {
    long dwords = 25;
    long cases = 80000;

    base[0] = 0;
    base[1] = 0;
    long pwr = alpha, acc = 0;
    for (long b = 2; b <= maxlen; b++) {
        acc += pwr;
        base[b] = acc;
        pwr *= alpha;
    }
    long tsize = acc + pwr;

    table = malloc((size_t)tsize);
    s = malloc((size_t)slen * sizeof(long));

    long state = 12345;
    long sink = 0;

    for (long c = 0; c < cases; c++) {
        for (long z = 0; z < tsize; z++) table[z] = 0;
        for (long i = 0; i < slen; i++) {
            state = (state * 1103515245 + 12345) & 2147483647;
            long r = state >> 16;
            s[i] = r % alpha;
        }
        for (long w = 0; w < dwords; w++) {
            state = (state * 1103515245 + 12345) & 2147483647;
            long rl = state >> 16;
            long span = maxlen - minlen + 1;
            long wlen = minlen + (rl % span);
            long code = 0;
            for (long k = 0; k < wlen; k++) {
                state = (state * 1103515245 + 12345) & 2147483647;
                long rc = state >> 16;
                code = code * alpha + (rc % alpha);
            }
            table[base[wlen] + code] = 1;
        }

        sink += count(0);
    }

    printf("%ld\n", sink);
    return 0;
}

File: leetcode/101-200/140-word-break-ii/bench/word_break_ii.rs
// Exponential-backtracking segmentation COUNT for LeetCode #140 (see
// word_break_ii.kara). Dict SET is a flat stamped base-A table.
const ALPHA: i64 = 3;
const MINLEN: i64 = 1;
const MAXLEN: i64 = 3;
const SLEN: i64 = 16;

fn count(s: &[i64], table: &[bool], base: &[i64], start: i64) -> i64 {
    if start == SLEN {
        return 1;
    }
    let mut total = 0i64;
    let mut code = 0i64;
    let mut end = start + 1;
    while end <= SLEN && end - start <= MAXLEN {
        code = code * ALPHA + s[(end - 1) as usize];
        let ln = end - start;
        if ln >= MINLEN {
            if table[(base[ln as usize] + code) as usize] {
                total += count(s, table, base, end);
            }
        }
        end += 1;
    }
    total
}

fn main() {
    let dwords: i64 = 25;
    let cases: i64 = 80000;

    let mut base: Vec<i64> = vec![0, 0];
    let mut pwr = ALPHA;
    let mut acc = 0i64;
    let mut b = 2i64;
    while b <= MAXLEN {
        acc += pwr;
        base.push(acc);
        pwr *= ALPHA;
        b += 1;
    }
    let tsize = acc + pwr;

    let mut table: Vec<bool> = vec![false; tsize as usize];
    let mut s: Vec<i64> = vec![0; SLEN as usize];

    let mut state: i64 = 12345;
    let mut sink: i64 = 0;

    for _ in 0..cases {
        for z in 0..tsize {
            table[z as usize] = false;
        }
        for i in 0..SLEN {
            state = (state * 1103515245 + 12345) & 2147483647;
            let r = state >> 16;
            s[i as usize] = r % ALPHA;
        }
        for _ in 0..dwords {
            state = (state * 1103515245 + 12345) & 2147483647;
            let rl = state >> 16;
            let span = MAXLEN - MINLEN + 1;
            let wlen = MINLEN + (rl % span);
            let mut code = 0i64;
            for _ in 0..wlen {
                state = (state * 1103515245 + 12345) & 2147483647;
                let rc = state >> 16;
                code = code * ALPHA + (rc % ALPHA);
            }
            table[(base[wlen as usize] + code) as usize] = true;
        }

        sink += count(&s, &table, &base, 0);
    }

    println!("{}", sink);
}
